File: packages/lite-filter/npm/src/Expression.ts
import type { Expression as IExpression, Adapter } from './types';
import { Clause } from './Clause';
import { Grouping, LogicalOperator } from './Grouping';
import { ComparisonOperator } from './ComparisonOperator';
import { parseFilter } from './parser';
import * as AdapterRegistry from './adapters';

/**
 * Expression - Main public API entry point
 * Provides factory methods for creating filter expressions
 */
export class Expression {
  // ========== Comparison Operations ==========

  static equals(property: string, value: any): IExpression {
    return new Clause(property, ComparisonOperator.EQUALS, value);
  }

  static notEquals(property: string, value: any): IExpression {
    return new Clause(property, ComparisonOperator.NOT_EQUALS, value);
  }

  static greaterThan(property: string, value: any): IExpression {
    return new Clause(property, ComparisonOperator.GREATER_THAN, value);
  }

  static greaterThanOrEqual(property: string, value: any): IExpression {
    return new Clause(property, ComparisonOperator.GREATER_THAN_OR_EQUAL, value);
  }

  static lessThan(property: string, value: any): IExpression {
    return new Clause(property, ComparisonOperator.LESS_THAN, value);
  }

  static lessThanOrEqual(property: string, value: any): IExpression {
    return new Clause(property, ComparisonOperator.LESS_THAN_OR_EQUAL, value);
  }

  static approxMatch(property: string, value: any): IExpression {
    return new Clause(property, ComparisonOperator.APPROX_MATCH, value);
  }

  static present(property: string): IExpression {
    return new Clause(property, ComparisonOperator.PRESENT, null);
  }

  // ========== String Operations ==========

  static contains(property: string, value: string): IExpression {
    return new Clause(property, ComparisonOperator.CONTAINS, value);
  }

  static startsWith(property: string, value: string): IExpression {
    return new Clause(property, ComparisonOperator.STARTS_WITH, value);
  }

  static endsWith(property: string, value: string): IExpression {
    return new Clause(property, ComparisonOperator.ENDS_WITH, value);
  }

  static matches(property: string, regex: string | RegExp): IExpression {
    return new Clause(property, ComparisonOperator.MATCHES, regex);
  }

  // ========== Null/Empty Checks ==========

  static isNull(property: string): IExpression {
    return new Clause(property, ComparisonOperator.IS_NULL, null);
  }

  static isEmpty(property: string): IExpression {
    return new Clause(property, ComparisonOperator.IS_EMPTY, null);
  }

  // ========== Array Operations ==========

  static includes(property: string, value: any): IExpression {
    return new Clause(property, ComparisonOperator.INCLUDES, value);
  }

  static includesAny(property: string, values: any[]): IExpression {
    return new Clause(property, ComparisonOperator.INCLUDES_ANY, values);
  }

  // ========== Numeric Operations ==========

  static between(property: string, min: number, max: number): IExpression {
    return new Clause(property, ComparisonOperator.BETWEEN, { min, max });
  }

  // ========== Date Operations ==========

  static withinDays(property: string, days: number): IExpression {
    return new Clause(property, ComparisonOperator.WITHIN_DAYS, days);
  }

  static year(property: string, year: number): IExpression {
    return new Clause(property, ComparisonOperator.YEAR, year);
  }

  // ========== Logical Operations ==========

  static and(...expressions: IExpression[]): IExpression {
    return new Grouping(LogicalOperator.AND, expressions);
  }

  static or(...expressions: IExpression[]): IExpression {
    return new Grouping(LogicalOperator.OR, expressions);
  }

  static not(expression: IExpression): IExpression {
    return new Grouping(LogicalOperator.NOT, [expression]);
  }

  // ========== Parser ==========

  /**
   * Parses an RFC4515 filter string into an Expression tree
   * @param filter RFC4515 filter syntax (e.g., "(&(status=active)(age>=18))")
   * @returns The parsed Expression
   * @throws Error if the filter syntax is invalid
   */
  static parse(filter: string): IExpression {
    return parseFilter(filter);
  }

  // ========== Adapter Management ==========

  /**
   * Registers an adapter for translating expressions to other query languages
   * @param key The adapter key (e.g., "SQL", "DynamoDB")
   * @param description Human-readable description
   * @param adapter The adapter implementation
   */
  static addAdapter(key: string, description: string, adapter: Adapter): void {
    AdapterRegistry.addAdapter(key, description, adapter);
  }

  /**
   * Lists all registered adapters
   * @returns Array of adapter keys and descriptions
   */
  static adapters(): Array<{ key: string; description: string }> {
    return AdapterRegistry.listAdapters();
  }

  /**
   * Gets an adapter by key (internal use)
   * @param key The adapter key
   * @returns The adapter
   * @throws Error if adapter not found
   */
  static getAdapter(key: string): Adapter {
    return AdapterRegistry.getAdapter(key);
  }
}

File: packages/lite-filter/java/src/main/java/com/zerobias/litefilter/Expression.java
package com.zerobias.litefilter;

import java.util.HashMap;
import java.util.Map;

/**
 * Expression interface for RFC4515 filter expressions.
 * Provides factory methods for building and parsing filter expressions.
 */
public interface Expression {

    /**
     * Test if an object matches this expression.
     *
     * @param obj the object to test
     * @return true if the object matches, false otherwise
     */
    boolean matches(Object obj);

    /**
     * Test if an object matches this expression with options.
     *
     * @param obj the object to test
     * @param options matching options
     * @return true if the object matches, false otherwise
     */
    boolean matches(Object obj, MatchOptions options);

    /**
     * Convert this expression using a registered adapter.
     *
     * @param adapterKey the key of the registered adapter
     * @return string representation in the target language
     * @throws IllegalArgumentException if adapter is not registered
     */
    String as(String adapterKey);

    // Adapter registry
    Map<String, AdapterInfo> ADAPTERS = new HashMap<>();

    /**
     * Register an adapter for converting expressions to other query languages.
     *
     * @param key the key to identify the adapter
     * @param description description of the adapter
     * @param adapter the adapter implementation
     */
    static void addAdapter(String key, String description, Adapter adapter) {
        ADAPTERS.put(key, new AdapterInfo(description, adapter));
    }

    /**
     * Get all registered adapters.
     *
     * @return map of adapter keys to descriptions
     */
    static Map<String, String> adapters() {
        Map<String, String> result = new HashMap<>();
        ADAPTERS.forEach((key, info) -> result.put(key, info.description));
        return result;
    }

    // Comparison operations

    /**
     * Create an equality expression.
     */
    static Expression equals(String property, Object value) {
        return new Clause(property, ComparisonOperator.EQUALS, value);
    }

    /**
     * Create a not equals expression.
     */
    static Expression notEquals(String property, Object value) {
        return new Clause(property, ComparisonOperator.NOT_EQUALS, value);
    }

    /**
     * Create a greater than expression.
     */
    static Expression greaterThan(String property, Object value) {
        return new Clause(property, ComparisonOperator.GREATER_THAN, value);
    }

    /**
     * Create a greater than or equal expression.
     */
    static Expression greaterThanOrEqual(String property, Object value) {
        return new Clause(property, ComparisonOperator.GREATER_THAN_OR_EQUAL, value);
    }

    /**
     * Create a less than expression.
     */
    static Expression lessThan(String property, Object value) {
        return new Clause(property, ComparisonOperator.LESS_THAN, value);
    }

    /**
     * Create a less than or equal expression.
     */
    static Expression lessThanOrEqual(String property, Object value) {
        return new Clause(property, ComparisonOperator.LESS_THAN_OR_EQUAL, value);
    }

    /**
     * Create an approximate match expression (fuzzy matching).
     */
    static Expression approxMatch(String property, String value) {
        return new Clause(property, ComparisonOperator.APPROX_MATCH, value);
    }

    /**
     * Create a presence check expression (not null/empty).
     */
    static Expression present(String property) {
        return new Clause(property, ComparisonOperator.PRESENCE_CHECK, null);
    }

    // String operations

    /**
     * Create a contains expression.
     */
    static Expression contains(String property, String value) {
        return new Clause(property, ComparisonOperator.CONTAINS, value);
    }

    /**
     * Create a starts with expression.
     */
    static Expression startsWith(String property, String value) {
        return new Clause(property, ComparisonOperator.BEGINS_WITH, value);
    }

    /**
     * Create an ends with expression.
     */
    static Expression endsWith(String property, String value) {
        return new Clause(property, ComparisonOperator.ENDS_WITH, value);
    }

    /**
     * Create a regex match expression.
     */
    static Expression matches(String property, String regex) {
        return new Clause(property, ComparisonOperator.REGEX, regex);
    }

    // Null/empty checks

    /**
     * Create an is null expression.
     */
    static Expression isNull(String property) {
        return new Clause(property, ComparisonOperator.IS_NULL, null);
    }

    /**
     * Create an is empty expression (for arrays).
     */
    static Expression isEmpty(String property) {
        return new Clause(property, ComparisonOperator.IS_EMPTY, null);
    }

    // Array operations

    /**
     * Create an includes expression (array contains element).
     */
    static Expression includes(String property, Object value) {
        return new Clause(property, ComparisonOperator.INCLUDES, value);
    }

    /**
     * Create an includes any expression (array intersects set).
     */
    static Expression includesAny(String property, String values) {
        return new Clause(property, ComparisonOperator.INCLUDES_ANY, values);
    }

    // Numeric operations

    /**
     * Create a between expression (inclusive range).
     */
    static Expression between(String property, Number min, Number max) {
        return new Clause(property, ComparisonOperator.BETWEEN, min + "," + max);
    }

    // Date operations

    /**
     * Create a within days expression (relative to now).
     */
    static Expression withinDays(String property, int days) {
        return new Clause(property, ComparisonOperator.WITHIN_DAYS, days);
    }

    /**
     * Create a year expression (extract year and compare).
     */
    static Expression year(String property, int year) {
        return new Clause(property, ComparisonOperator.YEAR, year);
    }

    // Logical operations

    /**
     * Create an AND expression.
     */
    static Expression and(Expression... expressions) {
        return new Grouping(LogicalOperator.AND, expressions);
    }

    /**
     * Create an OR expression.
     */
    static Expression or(Expression... expressions) {
        return new Grouping(LogicalOperator.OR, expressions);
    }

    /**
     * Create a NOT expression.
     */
    static Expression not(Expression expression) {
        return new Grouping(LogicalOperator.NOT, expression);
    }

    /**
     * Parse an RFC4515 filter string into an expression.
     *
     * @param filter the RFC4515 filter string
     * @return the parsed expression
     * @throws IllegalArgumentException if the filter is malformed
     */
    static Expression parse(String filter) {
        return RFC4515Parser.parse(filter);
    }

    /**
     * Internal class to hold adapter information.
     */
    class AdapterInfo {
        final String description;
        final Adapter adapter;

        AdapterInfo(String description, Adapter adapter) {
            this.description = description;
            this.adapter = adapter;
        }
    }
}
